core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala

```diff
@@ -16,131 +16,25 @@
 
-  def upload(bytes: Stream[F, Byte], bucket: Bucket, key: ObjectKey): Stream[F, Unit] =
-    Stream.bracketCase(startUpload(client, bucket, key)) {
-      case (createResponse, ExitCase.Canceled | ExitCase.Error(_)) =>
-        abortUpload(client, createResponse.uploadId, bucket, key).void
-      case (_, ExitCase.Completed) =>
-        Concurrent[F].unit
-    } flatMap (createResponse =>
-      for {
-        parts <- uploadParts(client, bytes, createResponse.uploadId, partSize, bucket, key)
-        _ <- Stream.eval(completeUpload(client, createResponse.uploadId, bucket, key, parts))
-      } yield ())
-
-  private def startUpload(
-    client: S3AsyncClient,
-    bucket: Bucket,
-    key: ObjectKey): F[CreateMultipartUploadResponse] =
-    Concurrent[F]
-      .delay(
-        client.createMultipartUpload(
-          CreateMultipartUploadRequest.builder.bucket(bucket.value).key(key.value).build))
-      .futureLift
-      .guarantee(ContextShift[F].shift)
-
-  private def uploadParts(
-    client: S3AsyncClient,
-    bytes: Stream[F, Byte],
-    uploadId: String,
-    minChunkSize: Int,
-    bucket: Bucket,
-    key: ObjectKey): Stream[F, List[CompletedPart]] =
-    (bytes.chunkMin(minChunkSize).zipWithIndex evalMap {
-      case (byteChunk, n) => {
-        // parts numbers must start at 1
-        val partNumber = Int.box(n.toInt + 1)
-
-        val uploadPartRequest =
-          UploadPartRequest.builder
-            .bucket(bucket.value)
-            .uploadId(uploadId)
-            .key(key.value)
-            .partNumber(partNumber)
-            .contentLength(Long.box(byteChunk.size.toLong))
-            .build
+  def upload(bytes: Stream[F, Byte], bucket: Bucket, key: BlobPath): Stream[F, Unit] = {
+    val service = S3PutService[F](client, partSize, bucket)
 
-        val uploadPartResponse =
-          Concurrent[F]
-            .delay(
-              client.uploadPart(
-                uploadPartRequest,
-                AsyncRequestBody.fromByteBuffer(byteChunk.toByteBuffer)))
-            .futureLift
-            .guarantee(ContextShift[F].shift)
-
-        uploadPartResponse.map(response =>
-          CompletedPart.builder.partNumber(partNumber).eTag(response.eTag).build)
-      }
-    }).foldMap(_ :: Nil)
-
-  private def completeUpload(
-    client: S3AsyncClient,
-    uploadId: String,
-    bucket: Bucket,
-    key: ObjectKey,
-    parts: List[CompletedPart]): F[CompleteMultipartUploadResponse] = {
-    val multipartUpload =
-      CompletedMultipartUpload.builder.parts(parts: _*).build
-
-    val completeMultipartUploadRequest =
-      CompleteMultipartUploadRequest.builder
-        .bucket(bucket.value)
-        .key(key.value)
-        .uploadId(uploadId)
-        .multipartUpload(multipartUpload)
-        .build
-
-    Concurrent[F]
-      .delay(client.completeMultipartUpload(completeMultipartUploadRequest))
-      .futureLift
-      .guarantee(ContextShift[F].shift)
+    Stream.eval(service((key, bytes))).void
```

Comment:
One downside of this change is that we `compile` the upload `Stream` in `async-blobstore` and then `Stream.eval` it here... 😕

Reply:
Curious, what's the reason this function returns `Stream[F, Unit]` instead of `F[Unit]`?